parcels/interaction/baseinteractionkernel.py

```diff
@@ -0,0 +1,145 @@
+import re
+import inspect
+from sys import version_info
+
+try:
+    from mpi4py import MPI
+except:
+    MPI = None
+
+from parcels.kernel.basekernel import BaseKernel
+
+__all__ = ['BaseKernel']
+
+
+re_indent = re.compile(r"^(\s+)")
+
+
+class BaseInteractionKernel(BaseKernel):
+    """Base super class for Interaction Kernel objects that encapsulates
+    auto-generated code.
+    """
+
+    def __init__(self, fieldset, ptype, pyfunc=None, funcname=None,
+                 funccode=None, py_ast=None, funcvars=None,
+                 c_include="", delete_cfiles=True):
+        if pyfunc is not None:
+            if isinstance(pyfunc, list):
+                funcname = ''.join([func.__name__ for func in pyfunc])
+            else:
+                funcname = pyfunc.__name__
+
+        super(BaseInteractionKernel, self).__init__(
+            fieldset=fieldset, ptype=ptype, pyfunc=pyfunc, funcname=funcname,
+            funccode=funccode, py_ast=py_ast, funcvars=funcvars,
+            c_include=c_include, delete_cfiles=delete_cfiles)
+
+        if pyfunc is not None:
+            if isinstance(pyfunc, list):
+                self._pyfunc = pyfunc
+            else:
+                self._pyfunc = [pyfunc]
+
+        # Generate the kernel function and add the outer loop
+        if self._ptype.uses_jit:
+            raise NotImplementedError("Interaction Kernels do not support"
+                                      " JIT mode currently.")
+
+    def __del__(self):
+        # Clean-up the in-memory dynamic linked libraries.
+        # This is not really necessary, as these programs are not that large, but with the new random
+        # naming scheme which is required on Windows OS'es to deal with updates to a Parcels' kernel.)
+        # It is particularly unneccessary for Interaction Kernels at this time (as this functionality
+        # is as of yet not implemented).
+        super().__del__()
+
+    @property
+    def ptype(self):
+        return self._ptype
+
+    @property
+    def pyfunc(self):
+        return self._pyfunc
+
+    @property
+    def fieldset(self):
+        return self._fieldset
+
+    @property
+    def c_include(self):
+        return self._c_include
```

Comment:
shouldn't those be part of the `BaseKernel`, or did you need to respecialize them due to the more complex logics implemented in their `BaseKernel` implementation ?

If it's not the latter and it's just the former, then it would be good to move them to the `BaseKernel` class.

Reply:
For the __init__ it needs a `BaseInteractionKernel`-specific part that makes sure the `_pyfunc` property contains a list of functions to be executed separately, rather than a consolidated single function as in normal kernels.
For the other functions I think you are right that they can be (and are) defined in the `BaseKernel` already. I will remove them.